roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala

```diff
@@ -517,66 +548,59 @@ object VirtualMachine {
       })
   }
 
-  def execute(op: OpApplyPrimArg, state: VMState): VMState =
-    state
-      .set(_ >> 'ctxt >> 'nargs)(op.nargs)
-      .updateSelf(state => {
-        val prim  = Prim.nthPrim(op.primNum)
-        val argno = op.arg
-
-        val (result, newState) =
-          if (op.unwind) { unwindAndApplyPrim(prim.get, state) } else {
-            // TODO: Fix
-            (prim.get.dispatchHelper(state.ctxt), state)
-          }
-
-        result match {
-          case Right(ob) =>
-            if (ob.is(Ob.OTsysval)) {
-              handleException(ob, op, newState.loc)
-              newState.set(_ >> 'doNextThreadFlag)(true)
-            } else if (argno >= newState.ctxt.argvec.elem.length) {
-              newState.set(_ >> 'vmErrorFlag)(true)
-            } else {
-              newState
-                .update(_ >> 'ctxt >> 'argvec >> 'elem)(_.updated(argno, ob))
-                .update(_ >> 'doNextThreadFlag)(if (op.next) true else _)
-            }
-
-          case Left(DeadThread) =>
-            newState.set(_ >> 'doNextThreadFlag)(true)
+  def execute(op: OpApplyPrimArg, state: VMState): VMState = {
+    val st = for {
+      _ <- modify(_.set(_ >> 'ctxt >> 'nargs)(op.nargs))
+
+      prim = Prim.nthPrim(op.primNum)
+      result <- runPrim(op.unwind, prim)
+
+      _ <- handlePrimResult(
+        result,
+        ob =>
+          State { vmState: VMState =>
+            if (ob.is(Ob.OTsysval))
+              //handleException(ob, op, vmState.loc)
+              (vmState.copy(doNextThreadFlag = true), ())
+            else if (op.arg >= vmState.ctxt.argvec.elem.size)
+              (vmState.copy(vmErrorFlag = true), ())
+            else
+              (vmState
+                 .update(_ >> 'ctxt >> 'argvec >> 'elem)(_.updated(op.arg, ob))
+                 .update(_ >> 'doNextThreadFlag)(if (op.next) true else _),
+               ())
         }
+      )
+    } yield ()
 
-      })
-
-  def execute(op: OpApplyPrimReg, state: VMState): VMState =
-    state
-      .set(_ >> 'ctxt >> 'nargs)(op.nargs)
-      .updateSelf(state => {
-        val prim  = Prim.nthPrim(op.primNum)
-        val regno = op.reg
+    st.runS(state).value
+  }
 
-        val (result, newState) =
-          if (op.unwind) { unwindAndApplyPrim(prim.get, state) } else {
-            // TODO: Fix
-            (prim.get.dispatchHelper(state.ctxt), state)
+  def execute(op: OpApplyPrimReg, state: VMState): VMState = {
```

Comment:
Personally, I would prefer to replace such signatures with `State[VMState,Unit]`. So, why don't we go for `State`?

Reply:
Yes, sure. But we would have to rewrite all `execute`s. That's something for another PR.